File: patterns/a9k40gel.go
package patterns

import (
	"time"

	"github.com/anderstorpsfestivalen/slisko/pkg/chassi"
	"github.com/anderstorpsfestivalen/slisko/pkg/configuration"
	"github.com/anderstorpsfestivalen/slisko/pkg/traffic"
)

type A9K40GE struct {
	ports []PortState
	style BlinkStyle
}

func (p *A9K40GE) Render(info RenderInfo, c *chassi.Chassi) {
	for i := range p.ports {
		p.ports[i].Render()
	}
}

func (p *A9K40GE) Info() PatternInfo {
	return PatternInfo{
		Name:     "a9k-40ge-l",
		Category: "misc",
	}
}

func (p *A9K40GE) Bootstrap(c *chassi.Chassi) {
	p.style = ASR9000Style()

	// Get or create default traffic shaper
	shaper := GetTrafficShaper()
	if shaper == nil {
		// Fallback to default if not set
		shaper = traffic.NewShaper(configuration.DefaultTrafficShaper())
	}

	for _, card := range c.GetCardOfType("A9K-40GE-L") {
		for _, port := range card.Link {
			time.Sleep(2 * time.Millisecond)
			p.ports = append(p.ports, p.style.CreatePort(port, shaper))
		}
	}
}

File: engine/src/patterns/asr9000/a9k_40ge_l.rs
use alloc::vec::Vec;

use super::blinkstyle::asr9000_style;
use crate::chassi::Chassi;
use crate::pattern::{BootstrapCtx, Pattern, PatternInfo, RenderInfo};
use crate::patterns::blinkstyle::PortState;

/// ASR9000 A9K-40GE-L linecard ports.
#[derive(Default)]
pub struct A9K40GE {
    ports: Vec<PortState>,
}

impl Pattern for A9K40GE {
    fn render(&mut self, info: &RenderInfo, c: &mut Chassi) {
        for port in &mut self.ports {
            port.render(&mut c.leds, info.traffic_millis);
        }
    }

    fn info(&self) -> PatternInfo {
        PatternInfo {
            name: "a9k-40ge-l",
            category: "misc",
        }
    }

    fn bootstrap(&mut self, c: &Chassi, ctx: &mut BootstrapCtx) {
        let style = asr9000_style();
        for index in c.link_indices_of_type("A9K-40GE-L") {
            self.ports.push(style.create_port(index, ctx));
        }
    }
}
